compiler/rustc_middle/src/middle/privacy.rs

```diff
@@ -27,26 +27,148 @@ pub enum AccessLevel {
     Public,
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, Debug, HashStable, Default)]
+pub struct EffectiveVisibility {
+    public: Option<Visibility>,
+    exported: Option<Visibility>,
+    reachable: Option<Visibility>,
+    reachable_from_impl_trait: Option<Visibility>,
+}
+
+impl EffectiveVisibility {
+    fn get_field(&self, tag: AccessLevel) -> &Option<Visibility> {
+        match tag {
+            AccessLevel::Public => &self.public,
+            AccessLevel::Exported => &self.exported,
+            AccessLevel::Reachable => &self.reachable,
+            AccessLevel::ReachableFromImplTrait => &self.reachable_from_impl_trait,
+        }
+    }
+
+    fn get_mut_field(&mut self, tag: AccessLevel) -> &mut Option<Visibility> {
+        match tag {
+            AccessLevel::Public => &mut self.public,
+            AccessLevel::Exported => &mut self.exported,
+            AccessLevel::Reachable => &mut self.reachable,
+            AccessLevel::ReachableFromImplTrait => &mut self.reachable_from_impl_trait,
+        }
+    }
+
+    pub fn get(&self, tag: AccessLevel) -> Option<&Visibility> {
+        (*self.get_field(tag)).as_ref()
+    }
+
+    fn get_mut(&mut self, tag: AccessLevel) -> Option<&mut Visibility> {
+        (*self.get_mut_field(tag)).as_mut()
+    }
+
+    fn set(&mut self, effective_vis: Option<Visibility>, tag: AccessLevel) {
+        *self.get_mut_field(tag) = effective_vis;
+    }
+
+    fn merge(
+        &mut self,
+        effective_vis_opt: Option<Visibility>,
+        tag: AccessLevel,
+        tree: impl DefIdTree,
+    ) {
+        let current_effective_vis = self.get_mut(tag);
+        if let Some(effective_vis) = effective_vis_opt {
+            if let Some(current_effective_vis) = current_effective_vis {
+                if current_effective_vis.is_at_least(effective_vis, tree) {
+                    *current_effective_vis = effective_vis;
+                }
+            } else {
+                self.set(effective_vis_opt, tag);
+            }
+        }
+    }
+
+    pub fn update(
+        &mut self,
+        effective_vis_opt: Option<Visibility>,
```

Comment:
```suggestion
        effective_vis: Option<Visibility>,
```
Style nit: the option-ness is only specified in the name when both `Foo` and `Option<Foo>` are in the same scope.
Same in `fn merge`.

Reply:
```suggestion
        effective_vis: Visibility,
```
Also this function is never called with `None` arguments.